include/data_sink.hpp

```diff
@@ -0,0 +1,900 @@
+                const auto view     = in_data.first(to_write);
+                pending_dataset->signal_values.insert(pending_dataset->signal_values.end(), view.begin(), view.end());
+
+                if (pending_dataset->signal_values.size() == maximum_window_size) {
+                    publish_dataset(std::move(*pending_dataset));
+                    pending_dataset.reset();
+                }
+            }
+        }
+
+        void
+        flush() override {
+            if (pending_dataset) {
+                publish_dataset(std::move(*pending_dataset));
+                pending_dataset.reset();
+            }
+            if (auto p = polling_handler.lock()) {
+                p->finished = true;
+            }
+        }
+    };
+
+    struct pending_snapshot {
+        property_map tag_data;
+        std::size_t  delay           = 0;
+        std::size_t  pending_samples = 0;
+    };
+
+    template<typename Callback, TriggerPredicate P>
+    struct snapshot_listener : public abstract_listener {
+        bool                          block = false;
+        std::chrono::nanoseconds      time_delay;
+        std::size_t                   sample_delay      = 0;
+        P                             trigger_predicate = {};
+        std::deque<pending_snapshot>  pending;
+        std::weak_ptr<dataset_poller> polling_handler = {};
+        Callback                      callback;
+
+        explicit snapshot_listener(P p, std::chrono::nanoseconds delay, std::shared_ptr<dataset_poller> poller, bool do_block)
+            : block(do_block), time_delay(delay), trigger_predicate(std::forward<P>(p)), polling_handler{ std::move(poller) } {}
+
+        explicit snapshot_listener(P p, std::chrono::nanoseconds delay, Callback cb) : trigger_predicate(std::forward<P>(p)), time_delay(std::forward<Callback>(cb)) {}
+
+        inline void
+        publish_dataset(DataSet<T> &&data) {
+            if constexpr (!std::is_same_v<Callback, null_type>) {
+                callback(std::move(data));
+            } else {
+                auto poller = polling_handler.lock();
+                if (!poller) {
+                    this->set_expired();
+                    return;
+                }
+
+                auto write_data = poller->writer.reserve_output_range(1);
+                if (block) {
+                    write_data[0] = std::move(data);
+                    write_data.publish(1);
+                } else {
+                    if (poller->writer.available() > 0) {
+                        write_data[0] = std::move(data);
+                        write_data.publish(1);
+                    } else {
+                        poller->drop_count++;
+                    }
+                }
+            }
+        }
+
+        void
+        set_sample_rate(float r) override {
+            sample_delay = std::round(std::chrono::duration_cast<std::chrono::duration<float>>(time_delay).count() * r);
```

Comment:
Naming: 'set_sample_rate' when setting a 'sampling_delay'?

'r' = ??? -> might be useful to define this as a std::chrono::duration type.

Reply:
it's calculating the delay from the rate (`r`). But yes, naming could be improved.